### PR TITLE
Fix the server side socket bug & leader's board init & update & final state

### PR DESCRIPTION
**Pull request by ZiqiXu404**

### CODE REVIEW

File: src/modules/scene/name_scene.py [hunk 1]
```diff
@@ -26,7 +31,8 @@ def start_scene(self):
 
                 if event.type == pygame.KEYDOWN:
                     if event.key == pygame.K_RETURN:
-                        # TODO(nickbar01234) - Send to server
+                        self.get_network().send_name(name)
+                        self.__network_barrier.release()
```
Comment (ZiqiXu404) [inline src/modules/scene/name_scene.py:35]: critical change - this fixes a bug where sometimes the name acknowledge/response will arive out of order and cause InvalidMessage exception thrown

Comment (nickbar01234) [reply src/modules/scene/name_scene.py:35]: Good catch!

File: src/server.py [hunk 2]
```diff
@@ -99,26 +108,47 @@ def player_listener(self, player_socket: socket.socket, player_addr):
...
+    # i.e., no other messages should arrive / be sent during the broadcast
+    #
+    def broadcast_with_ack(self, summary, encoded_message):
+        player_sockets = self.__state.get_all_player_sockets_with_locks()
+        for name, player_socket, lock in player_sockets:
+            with lock:
+                player_socket.sendall(encoded_message)
+        for name, player_socket, lock in player_sockets:
+            with lock:
+                _ack = s.decode_ack(player_socket.recv(1024))  # Receiving ack
```
Comment (nickbar01234) [inline src/server.py:131]: Wouldn't this run into problems if `player_listener` grab the thread midway?

**Review by nickbar01234 — APPROVED** [review]
LGTM! Besides `broadcast_with_ack`, I think everything else looks good!

**Pull request merged**